hrl/agent/dsc/classifier/sift_classifier.py

```diff
@@ -0,0 +1,322 @@
+        optimistic_classifier = BOVWClassifier(num_clusters=num_clusters, num_sift_keypoints=num_sift_keypoints)
+        self.pessimistic_classifier_for_pos_data = BOVWClassifier(num_clusters=num_clusters, num_sift_keypoints=num_sift_keypoints)
+        self.pessimistic_classifier_for_neg_data = BOVWClassifier(num_clusters=num_clusters, num_sift_keypoints=num_sift_keypoints)
+
+        self.gamma = gamma
+        self.nu = nu
+
+        self.positive_examples = deque([], maxlen=buffer_size)
+        self.negative_examples = deque([], maxlen=buffer_size)
+        
+        super().__init__(optimistic_classifier, self.pessimistic_classifier_for_pos_data)
+    
+    def is_initialized(self):
+        return (self.optimistic_classifier.is_initialized() 
+            and self.pessimistic_classifier_for_pos_data.is_initialized()
+            and self.pessimistic_classifier_for_neg_data.is_initialized())
+
+    def optimistic_predict(self, state):
+        assert isinstance(state, np.ndarray)
+        return self.optimistic_classifier.predict([state])[0] == 1
+
+    def pessimistic_predict(self, state):
+        assert isinstance(state, np.ndarray)
+        assert state.shape == (84, 84)
+        return self.pessimistic_classifier_for_pos_data.predict([state])[0] == 1 and self.pessimistic_classifier_for_neg_data.predict([state])[0] == -1
+
+    def add_positive_examples(self, images, positions):
+        assert len(images) == len(positions)
+
+        positive_examples = [TrainingExample(img, pos) for img, pos in zip(images, positions)]
+        self.positive_examples.append(positive_examples)
+
+    def add_negative_examples(self, images, positions):
+        assert len(images) == len(positions)
+
+        negative_examples = [TrainingExample(img, pos) for img, pos in zip(images, positions)]
+        self.negative_examples.append(negative_examples)
+
+    @staticmethod
+    def extract_positions(examples):
+        examples = itertools.chain.from_iterable(examples)
+        positions = [example.pos for example in examples]
+        return np.array(positions)
+
+    @staticmethod
+    def construct_image_list(examples):
+        """
+        examples is either self.positive_examples or self.nagetive_examples
+        """
+        images = [example.obs._frames[-1].squeeze() for trajectory in examples for example in trajectory]
+        return images
+
+    def fit_initiation_classifier(self):
+        if len(self.negative_examples) > 0 and len(self.positive_examples) > 0:
+            self.train_two_class_classifier()
+        elif len(self.positive_examples) > 0:
+            self.train_one_class_classifier()
+
+    def train_one_class_classifier(self):
+        # data
+        positive_examples = self.construct_image_list(self.positive_examples)
+        positive_labels = [1 for _ in positive_examples]
+        # pessimistic
+        self.pessimistic_classifier.fit(positive_examples, positive_labels, svm_type='one_class_svm', nu=0.1)
```

Comment:
Here we are only fitting the optimistic classifier for the positive data. What about the optimistic classifier for the negative data?